### PR TITLE
Mpuncel/pod-factory

### PR DESCRIPTION
**Pull request by mpuncel**

### CODE REVIEW

**Comment by bwester** [comment]
Sorry, my earlier PR also changes the Pod type, hence the conflicts.

File: bin/p2-restart/main.go [hunk 1]
```diff
@@ -24,8 +24,8 @@ $ p2-restart --pod-dir /custom/pod/home
 
 `)
 
-	podName  = restart.Arg("pod-name", fmt.Sprintf("The name of the pod to be restarted. Looks in the default pod home '%s' for the pod", pods.DEFAULT_PATH)).String()
 	nodeName = restart.Flag("node-name", "The name of this node (default: hostname)").String()
+	podName  = restart.Arg("pod-name", fmt.Sprintf("The name of the pod to be restarted. Looks in the default pod home '%s' for the pod", pods.DefaultPath)).String()
```
Comment (bwester) [inline bin/p2-restart/main.go]: nit: Could you keep the two `Flag` definitions together?

**Comment by mpuncel** [comment]
rebased on master

File: pkg/preparer/orchestrate_test.go [hunk 2]
```diff
@@ -85,7 +85,7 @@ func (t *TestPod) Node() types.NodeName {
 	return "hostname"
 }
 
-func (t *TestPod) Path() string {
+func (t *TestPod) Home() string {
```
Comment (petertseng) [inline pkg/preparer/orchestrate_test.go]: I have a suspicion this would belong in the other commit

File: bin/p2-launch/main.go [hunk 3]
```diff
@@ -55,7 +55,8 @@ func main() {
 		log.Fatalf("%s", err)
 	}
 
-	pod := pods.NewPod(manifest.ID(), types.NodeName(*nodeName), pods.PodPath(*podRoot, manifest.ID()))
+	podFactory := pods.NewFactory(*podRoot, types.NodeName(*nodeName))
+	pod := podFactory.NewPod(manifest.ID())
```
Comment (petertseng) [inline bin/p2-launch/main.go]: I could see this being one line to make it clear this is the only use of this factory. unless you think there will be more later. I wouldn't insist on this before shipping though.

File: pkg/hooks/hooks_test.go [hunk 4]
```diff
@@ -27,8 +27,13 @@ func TestExecutableHooksAreRun(t *testing.T) {
 
 	ioutil.WriteFile(path.Join(tempDir, "test1"), []byte("#!/bin/sh\necho $HOOKED_POD_ID > $(dirname $0)/output"), 0755)
 
+	// So PodFromPodHome doesn't bail out, write a minimal current_manifest.yaml
+	ioutil.WriteFile(path.Join(podDir, "current_manifest.yaml"), []byte("id: my_hook"), 0755)
```
Comment (petertseng) [inline pkg/hooks/hooks_test.go]: and will this get removed after the test?

answered own question. yes. entire podDir gets removed.

Comment (petertseng) [inline pkg/hooks/hooks_test.go]: if this becomes too painful (I do see it's repeated three times) I might suggest putting it in a function

File: pkg/pods/pod_test.go [hunk 5]
```diff
@@ -29,7 +28,9 @@ import (
 )
 
 func getTestPod() *Pod {
-	return NewPod(types.PodID("hello"), "testNode", "/data/pods/test")
+	// Can't use public method because pod home is different than pod id
```
Comment (petertseng) [inline pkg/pods/pod_test.go]: what is the public method that you can't use? I only see exported methods being used here, so I think I don't understand the comment

File: pkg/pods/pod_test.go [hunk 6]
```diff
@@ -136,7 +137,8 @@ config:
 
 	podTemp, _ := ioutil.TempDir("", "pod")
 
-	pod := NewPod(manifest.ID(), "testNode", PodPath(podTemp, manifest.ID()))
+	podFactory := NewFactory(podTemp, "testNode")
+	pod := podFactory.NewPod(manifest.ID())
```
Comment (petertseng) [inline pkg/pods/pod_test.go]: I can see all these potentially being a single line as well, but I don't care

File: pkg/pods/pod_test.go (hunk 5 above)
Comment (mpuncel) [inline pkg/pods/pod_test.go]: i don't recall what I meant by that so i'll delete it lol

**Comment by petertseng** [comment]
I think I've said all that needs to be said 👍

**Pull request merged**